samples/dynamic_shapes/pytorch_dynamic_shapes.ipynb

```diff
@@ -0,0 +1,560 @@
+            "    %c1_i32 = arith.constant 1 : i32\n",
+            "    %dim = tensor.dim %arg0, %c0 : tensor<?xi32>\n",
+            "    %0 = tensor.empty(%dim) : tensor<?xi32>\n",
+            "    %1 = linalg.generic {indexing_maps = [#map, #map], iterator_types = [\"parallel\"]} ins(%arg0 : tensor<?xi32>) outs(%0 : tensor<?xi32>) {\n",
+            "    ^bb0(%in: i32, %out: i32):\n",
+            "      %2 = arith.addi %in, %c1_i32 : i32\n",
+            "      linalg.yield %2 : i32\n",
+            "    } -> tensor<?xi32>\n",
+            "    return %1 : tensor<?xi32>\n",
+            "  }\n",
+            "}\n"
+          ]
+        }
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "## Test the imported program\n",
+        "\n",
+        "_Note: you can stop after each step and use intermediate outputs with other tools outside of Colab._\n",
+        "\n",
+        "_See the [README](https://github.com/openxla/iree/tree/main/samples/dynamic_shapes#instructions) for more details and example command line instructions._\n",
+        "\n",
+        "* _The \"imported MLIR\" (above) can be used by IREE's generic compiler tools_\n",
+        "* _The \"binary\" can be saved and used by runtime applications_\n",
+        "\n",
+        "_The specific point at which you switch from Python to native tools will depend on your project._"
+      ],
+      "metadata": {
+        "id": "z6w_Pbl6tUtJ"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "# Export and compile.\n",
+        "exported_output = aot.export(DynamicShapesModule)\n",
+        "\n",
+        "# Compile to a file on disk for usage outside of Python.\n",
+        "flatbuffer_path = os.path.join(ARTIFACTS_DIR, \"dynamic_shapes_cpu.vmfb\")\n",
+        "exported_output.compile(save_to=flatbuffer_path)\n",
+        "print(f\"Wrote compiled program to path '{flatbuffer_path}'\")\n",
+        "\n",
+        "# Compile into memory for testing.\n",
+        "binary = exported_output.compile(save_to=None)"
+      ],
+      "metadata": {
+        "colab": {
+          "base_uri": "https://localhost:8080/"
+        },
+        "id": "0PGyH1tvI_Ic",
+        "outputId": "23b53928-4d77-461f-e4b8-b2c8ffb25ef0"
+      },
+      "execution_count": 7,
+      "outputs": [
+        {
+          "output_type": "stream",
+          "name": "stdout",
+          "text": [
+            "Wrote compiled program to path '/tmp/iree/colab_artifacts/dynamic_shapes_cpu.vmfb'\n"
+          ]
+        }
+      ]
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "import iree.runtime as ireert\n",
+        "import numpy as np\n",
+        "\n",
+        "# Use the IREE runtime API to test the compiled program.\n",
+        "config = ireert.Config(\"local-task\")\n",
+        "vm_module = ireert.load_vm_module(\n",
+        "    ireert.VmModule.wrap_buffer(config.vm_instance, binary.map_memory()),\n",
```

Comment:
The last time I've typed this out, I wasn't happy with some of the rough edges. I think I'd like some overloads to:

* Perhaps automatically handle `binary` without needing to `map_memory()`
* Add a method like `create_module` to `VmInstance`
* Have a look at `ireert.load_vm_module`... never been thrilled with that and feel like there is a better way.

Basically, would be nice if canonical usage was more of a one-liner rooted on an object instead of multiple top-level types/exports.

Reply:
Agreed, a canonical one-liner would be nice. This is a little bit shorter than what the TF notebook has, at least... but it still could be condensed.

```python
# This doesn't feel like Python, maybe just a naming thing though
flatbuffer_blob = compile_str(compiler_module, target_backends=["llvm-cpu"], input_type="stablehlo")

# Lots of setup for "load the program"
config = ireert.Config("local-task")
ctx = ireert.SystemContext(config=config)
vm_module = ireert.VmModule.from_flatbuffer(ctx.instance, flatbuffer_blob)
ctx.add_vm_module(vm_module)
```

Reply:
I mean, this code dates back to the very first demo... Due for a scrub